Switch method of matching post types to models

mkaz opened the pull request:
## Description

This change shows a different way of mapping post type to the wp.api.models. I'm including here for a discussion, since I'm not sure why the current method is in place and if there is something I'm missing.

The current method, including [core change #41111](https://core.trac.wordpress.org/ticket/41111) constructs API routes with regular expressions and then tries to determine the model based on those routes. The routes are then compared against each other as strings, and not tested as regular expressions.

This is troublesome because the API routes can be different per install, for example I'm working on bringing Gutenberg to WordPress.com which uses an external API server, and includes a `sites/BLOG_ID/` parameter in the routes. So 

The routes are also based off the wp.api.versionString which can be changed, and was recently in  #3377 as [you can see here](https://github.com/WordPress/gutenberg/blob/master/lib/client-assets.php#L793) it was changed from `wp/v2` to `gutenberg/v1` which I don't think is a valid API route. The only reason I think this doesn't break the mapping function is written to the page source prior to the versionString change.

This change does away with the routes mapping, and simple maps the types to objects that need to be mapped, in this case it is primarily `attachment => media`

I also added a filter which could be used to extend the mapping or override any that is necessary.

Additionally this change should make it much quicker, since the _.find has to traverse the prototype for each object which in this case is quite deep to find and check the route.

mkaz committed:
Switch method of matching types to models

mkaz committed:
lint fixes

aduth commented on lib/client-assets.php:
```diff
@@ -440,39 +440,39 @@ function gutenberg_register_vendor_script( $handle, $src, $deps = array() ) {
-		$post_type_rest_base_mapping[ $post_type_object->name ] = $rest_base;
-	}
+	$post_type_rest_base_mapping               = array();
+	$post_type_rest_base_mapping['attachment'] = 'media';
```
Why was this changed from the previous implementation where we use the post type objects to determine the `rest_base`? I don't think we want to hard-code attachment here, especially since we should respect the `rest_base` of any post type, without requiring that the plugin author opt in to `post_type_rest_base_mapping` (questioning whether the filter is even necessary, or redundant with `rest_base`).

mkaz replied:
The original implementation changed the mappings for post => posts, and page => pages both where the model object is not plural. I think the rest base and route would be specified in the model, so I don't think we need it for determining the model.

aduth replied:
>The original implementation changed the mappings for post => posts, and page => pages both where the model object is not plural.

But this is handled in the client. Even in your current implementation you're passing the mapping in as the plural.

Handled here:

https://github.com/WordPress/wordpress-develop/blob/1c2b0762c7f1a91394c633ae812505919bd297ed/src/wp-includes/js/wp-api.js#L1190-L1223

I'm not really sure why such a mapping exists, but I don't know that we need to worry so much about singular-ifying custom types?

Even if we were to create a mapping, we'd probably want to take advantage of the existing `wpApiSettings.mapping` rather than creating a separate variable.

cc @adamsilverstein

mkaz commented:
Switching method to `wp.api.getModelByType` which will better reflect the function. See [Core Ticket #42737](https://core.trac.wordpress.org/ticket/42737)

mkaz committed:
Switch methods to look up model by type

This also simplifies by combining the taxonomy model look up and
post model look up into the same function, since they were doing
the same thing of looking up a model.

Additional the taxonomy collection was renamed to getCollectionByType
which more accurately reflects the function behavior and can be
used to look up any collection by type.

mkaz committed:
Switch JS calls to use new getCollectionByType functions

mkaz committed:
Switch model calls to use new getModelByType functions

mkaz commented:
I don't really like the mapping of values by hand [done here](https://github.com/WordPress/gutenberg/pull/3683/commits/d77f4e0768142f72207b22e07caf8206660e38e8#diff-620130744b6b73b822a3df609671d930R445), so welcome any suggestions. 

There are so few that have different names than the models, I'm not sure where else that mapping might already exist either in WP or the API.

BE-Webdesign commented on lib/client-assets.php:
```diff
@@ -432,47 +432,52 @@ function gutenberg_register_vendor_script( $handle, $src, $deps = array() ) {
+	$type_model_mapping               = array();
+	$type_model_mapping['attachment'] = 'media';
+	$type_model_mapping['post_tag']   = 'tag';
+	// You can use the filter to extend or map a custom type to another.
```
Not sure if I am following, but maybe the above could be:

```php
$type_model_mapping = array(
  'attachment' => 'media',
  'post_tag'   => 'tag',
);
```

Same for `$type_collection_mapping`?

BE-Webdesign replied:
> I don't really like the mapping of values by hand done here, so welcome any suggestions.

I think manual is the way to go, it has to be coded somewhere. Maybe there is something in the `WP_Post_Type` object that already contains this information, and we could use that to auto-magically map everything?

mkaz committed:
Simplify defining array

mkaz commented:
@BE-Webdesign Thank you for taking a look, and the suggestion. You're right defining the array that way is much better.

I think I started the more verbose way of because I was going to have it empty, but then changed my mind and realized setting the defaults made more sense.

All cleaned up in e2d1082

BE-Webdesign commented:
Could we use `rest_base` to generate these mappings? I am still foggy on what this is doing exactly, so hopefully someone else can come along and review this better 😄.

aduth commented:
In case it was overlooked, noting that I dropped a comment at https://github.com/WordPress/gutenberg/pull/3683#discussion_r155082207

mkaz commented:
@youknowriad PR https://github.com/WordPress/gutenberg/pull/5253 removed the necessity to map into Backbone models. This is no longer needed.
